flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerService.java

```diff
@@ -18,43 +18,306 @@
+	 * {@link org.apache.flink.streaming.api.watermark.Watermark Watermark}.
+	 */
+	private long currentWatermark = Long.MIN_VALUE;
+
+	// Variables to be set when the service is started.
+
+	protected TypeSerializer<K> keySerializer;
+
+	protected TypeSerializer<N> namespaceSerializer;
+
+	private InternalTimer.TimerSerializer<K, N> timerSerializer;
+
+	protected Triggerable<K, N> triggerTarget;
+
+	private volatile boolean isInitialized;
+
+	public InternalTimerService(
+			int totalKeyGroups, 
+			KeyGroupRange keyGroupRange, 
+			KeyContext keyContext, 
+			ProcessingTimeService processingTimeService) {
+		
+		this.totalKeyGroups = totalKeyGroups;
+		this.keyGroupRange = checkNotNull(keyGroupRange);
+		this.keyContext = checkNotNull(keyContext);
+		this.processingTimeService = checkNotNull(processingTimeService);
+	}
 
 	/** Returns the current processing time. */
-	long currentProcessingTime();
+	public long currentProcessingTime() {
+		return processingTimeService.getCurrentProcessingTime();
+	}
 
 	/** Returns the current event-time watermark. */
-	long currentWatermark();
+	public long currentWatermark() {
+		return currentWatermark;
+	}
 
 	/**
 	 * Registers a timer to be fired when processing time passes the given time. The namespace
 	 * you pass here will be provided when the timer fires.
 	 */
-	void registerProcessingTimeTimer(N namespace, long time);
+	abstract public void registerProcessingTimeTimer(N namespace, long time);
 
 	/**
 	 * Deletes the timer for the given key and namespace.
 	 */
-	void deleteProcessingTimeTimer(N namespace, long time);
+	abstract public void deleteProcessingTimeTimer(N namespace, long time);
 
 	/**
 	 * Registers a timer to be fired when processing time passes the given time. The namespace
 	 * you pass here will be provided when the timer fires.
 	 */
-	void registerEventTimeTimer(N namespace, long time);
+	abstract public void registerEventTimeTimer(N namespace, long time);
 
 	/**
 	 * Deletes the timer for the given key and namespace.
 	 */
-	void deleteEventTimeTimer(N namespace, long time);
+	abstract public void deleteEventTimeTimer(N namespace, long time);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getEventTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Returns the timers for the given key group.
+	 */
+	abstract public Set<InternalTimer<K, N>> getProcessingTimeTimersForKeyGroup(int keyGroup);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreEventTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Restores the timers for the given key group.
+	 */
+	abstract public void restoreProcessingTimeTimersForKeyGroup(int keyGroup, Iterable<InternalTimer<K, N>> timers);
+
+	/**
+	 * Starts the execution of the timer service
+	 */
+	abstract public void start();
+
+	/**
+	 * Closes the timer service.
+	 */
+	abstract public void close();
```

Comment:
I would suggest to let the class implement `Closeable` and have the `throws IOException` in the signature. E.g. the implementation for RocksDB would even like to report `IOException`s.